js/mirador_viewer.js

```diff
@@ -12,10 +12,13 @@
      */
     Drupal.behaviors.Mirador = {
         attach: function (context, settings) {
+            Drupal.IslandoraMirador.instances = Drupal.IslandoraMirador.instances  || {}
             Object.entries(settings.mirador.viewers).forEach(entry => {
               const [base, values] = entry;
               once('mirador-viewer', base, context).forEach(() =>
-                Mirador.viewer(values, window.miradorPlugins || {})
+                // save the mirador instance so other modules can interact
+                // with the store/actions at e.g. Drupal.IslandoraMirador.instances["#mirador-xyz"].store
+                Drupal.IslandoraMirador.instances[viewerId] = Mirador.viewer(values, window.miradorPlugins || {})
```

Comment:
Originally commented as https://github.com/Islandora/islandora_mirador/pull/47#discussion_r1654924333 , but yeah:

Also, should probably have something on the `detach` side, to avoid potentially leaking instances. Would probably have to be conditional on the return from `once.remove()`?

Reply:
Really, should possibly move the definition:

```js
Drupal.IslandoraMirador.instances = Drupal.IslandoraMirador.instances  || {}
```

outside of the definition of the behaviors entirely..

Reply:
> Also, should probably have something on the detach side, to avoid potentially leaking instances. Would probably have to be conditional on the return from once.remove()?

I'm not sure I follow. How would this issue/concern ever surface if we don't handle in once.remove?

Reply:
If there was ever a variables number of viewers in the page, say, with AJAX adding or removing, or perhaps some flavour of endless-scrolling kind of things, then the instances for viewers removed from the page would remained tracked in our new variable.

Reply:
Thanks - I guess I could see that happening.

> outside of the definition of the behaviors entirely.

Where would we put this then?

Reply:
Though thinking more about once.remove... Wouldn't the object be still sitting in memory in the DOM with the code as it stands today? We're just adding a hash/map here that points to the object

Reply:
Javascript is a garbage-collected language, and we're creating/persisting a reference that would prevent garbage collection of the value in question. I'm not intimately familiar with the internals of Mirador, but I'm assuming it either binds to elements (as in, those which are being removed, so the reference already goes away when the elements are removed from the page) or somewhere in the document context (in which case we should possibly do more to tear it down?).

On the other hand, if there is already some reference in the DOM (as it seems there must be, given the viewer JS continues running) then what is the utility in our adding another? A bit of poking around in Mirador, finding https://github.com/ProjectMirador/mirador/blob/573444629ce8fec1d1fc40c9617231b716cd6140/src/lib/MiradorViewer.js#L21-L26 , which leads me to go inspect the DOM properties of the viewer element in the page, and indeed, there appear to be some react stuff on the element itself... so it would should go away when the element is removed from the page, leaving only the instance we track.

Reply:
I'm just going off docs and code examples - I don't know the mirador well at all. But IIUC the element is somewhere in the DOM and the reference in memory is only returned in when creating the viewer object on a DOM element. I don't know how else to get to it to access miradors' state/dispatch calls.

I'll just add a `delete` call in detach.